### PR TITLE
ERMS-2317: Titelverlängerung: Titelauswahl umstrukturieren

### PR DESCRIPTION
**Pull request by Moetez16**

### CODE REVIEW

File: app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy [hunk 1]
```diff
@@ -1112,29 +1112,44 @@ class SubscriptionController extends AbstractDebugController {
     @DebugAnnotation(test = 'hasAffiliation("INST_USER")')
     @Secured(closure = { ctx.springSecurityService.getCurrentUser()?.hasAffiliation("INST_USER") })
     def renewEntitlementsWithSurvey() {
+        Map<String, Object> result = [:]
+        result.institution = contextService.getOrg()
+        result.user = User.get(springSecurityService.principal.id)
         params.id = params.targetSubscriptionId
         params.sourceSubscriptionId = Subscription.get(params.targetSubscriptionId)?.instanceOf?.id
+        result.max = params.max ? Integer.parseInt(params.max) : result.user.getDefaultPageSizeTMP().toInteger()
```
Comment (klodav) [inline app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy:1120]: Better change return value of User.getDefaultPageSizeTMP()

Comment (croqueGrec09) [reply app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy:1120]: Agree; but that may cause collateral damages!

File: app/CHANGELOG-TMP.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 **Ticket    Date    Branch      Author  Feature/Bug     Description/Keywords**
 
+2317    02.01.2020  rc1.3       Moe     Feature     Titelverlängerung: Titelauswahl umstrukturieren
```
Comment (croqueGrec09) [inline app/CHANGELOG-TMP.md:5]: Yeah yeah, January 2nd ... :D

File: app/grails-app/views/subscription/renewEntitlements.gsp [hunk 3]
```diff
@@ -98,15 +98,15 @@ ${message(code: 'subscription.details.availableTitles')} ( ${message(code: 'defa
...
+                if (~delIdx) tippsToAdd.splice(delIdx,1);
                 $("tr[data-index='"+index+"'").removeClass("positive").addClass("negative");
                 corresp.find(".bulkcheck:checked").prop("checked", false);
                 tippsToDelete.push($(this).parents("tr").attr("data-gokbId"));
             }
             else {
                 $("tr[data-index='"+index+"'").removeClass("negative");
                 var delIdx = tippsToDelete.indexOf($(this).parents("tr").attr("data-gokbId"));
-                if (~delIdx) tippsToDelete.slice(delIdx,1);
+                if (~delIdx) tippsToDelete.slpice(delIdx,1);
```
Comment (croqueGrec09) [inline app/grails-app/views/subscription/renewEntitlements.gsp:109]: typo

File: app/grails-app/views/subscription/showEntitlementsRenewWithSurvey.gsp [hunk 4]
```diff
@@ -91,7 +91,7 @@
                     <td class="titleCell">
                         <semui:ieAcceptStatusIcon status="${ie?.acceptStatus}"/>
 
-                        <semui:listIcon type="${tipp.title?.type?.value}"/>
+                        <semui:listIcon type="${tipp.title?.class.name}"/>
```
Comment (croqueGrec09) [inline app/grails-app/views/subscription/showEntitlementsRenewWithSurvey.gsp:94]: There is a method title.printTitleType() for that.

File: app/grails-app/views/package/compare.gsp [hunk 5]
```diff
@@ -208,7 +208,7 @@
 		<tr>
 			
 			<td>
-				<semui:listIcon type="${currentTitle[0]?.type?.value}"/><strong><g:link action="show" controller="title" id="${currentTitle[0].id}">${entry.key}</g:link></strong>
+				<semui:listIcon type="${currentTitle[0]?.medium?.value}"/><strong><g:link action="show" controller="title" id="${currentTitle[0].id}">${entry.key}</g:link></strong>
```
Comment (klodav) [inline app/grails-app/views/package/compare.gsp:211]: ?

File: app/grails-app/views/public/gascoDetails.gsp [hunk 6]
```diff
@@ -68,7 +68,7 @@
                 <tr>
                     <td>${counter + 1}</td>
                     <td>
-                        <semui:listIcon type="${tipp?.title?.type?.value}"/>
+                        <semui:listIcon type="${tipp?.title?.medium?.value}"/>
```
Comment (klodav) [inline app/grails-app/views/public/gascoDetails.gsp:71]: ?

File: app/grails-app/views/survey/_actions.gsp [hunk 7]
```diff
@@ -53,7 +53,7 @@
             </g:if>
 
             <g:if test="${surveyInfo.isSubscriptionSurvey && surveyConfig && surveyConfig?.type == 'Subscription' && !surveyConfig?.pickAndChoose
-                    && surveyInfo.status?.id in [de.laser.helper.RDStore.SURVEY_IN_EVALUATION?.id, de.laser.helper.RDStore.SURVEY_COMPLETED?.id]}">
+                    && surveyInfo.status?.id in [de.laser.helper.RDStore.SURVEY_SURVEY_COMPLETED?.id, de.laser.helper.RDStore.SURVEY_IN_EVALUATION?.id, de.laser.helper.RDStore.SURVEY_COMPLETED?.id]}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_actions.gsp:56]: no need of ?, they are systemwide constants

File: app/grails-app/views/public/gascoDetailsIssueEntitlements.gsp [hunk 8]
```diff
@@ -68,7 +68,7 @@
                 <tr>
                     <td>${counter + 1}</td>
                     <td>
-                        <semui:listIcon type="${tipp.title?.type?.value}"/>
+                        <semui:listIcon type="${tipp.title?.medium?.value}"/>
```
Comment (klodav) [inline app/grails-app/views/public/gascoDetailsIssueEntitlements.gsp:71]: ?

File: app/grails-app/views/survey/_actions.gsp [hunk 9]
```diff
@@ -79,6 +79,12 @@
 
             <g:if test="${surveyInfo && surveyInfo.status?.id in [de.laser.helper.RDStore.SURVEY_IN_EVALUATION?.id, de.laser.helper.RDStore.SURVEY_SURVEY_COMPLETED?.id,de.laser.helper.RDStore.SURVEY_COMPLETED?.id]}">
 
+                <g:if test="${surveyInfo && surveyInfo.status?.id == de.laser.helper.RDStore.SURVEY_IN_EVALUATION?.id}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_actions.gsp:82]: to be or not to be => ?

File: app/grails-app/views/subscription/compare.gsp [hunk 10]
```diff
@@ -190,7 +190,7 @@
 							<g:set var="highlight" value="${entry.value[2]}"/>
 							<tr>
 								
-								<td><semui:listIcon type="${currentTitle?.type?.value}"/>
+								<td><semui:listIcon type="${currentTitle?.medium?.value}"/>
```
Comment (klodav) [inline app/grails-app/views/subscription/compare.gsp:193]: ?

File: app/grails-app/views/survey/surveyParticipants.gsp [hunk 11]
```diff
@@ -50,33 +50,35 @@
...
 <br>
 
-<g:if test="${surveyConfigs}">
+<g:if test="${surveyConfig}">
     <div class="ui grid">
         <div class="sixteen wide stretched column">
             <div class="ui top attached tabular menu">
 
-                <g:if test="${surveyConfig?.type == 'Subscription'}">
+                <g:if test="${surveyConfig.type == 'Subscription' || surveyConfig.type == 'IssueEntitlementsSurvey'}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/surveyParticipants.gsp:58]: I suggest for later to define surveyConfig.type constants

File: app/grails-app/views/survey/surveyTitles.gsp [hunk 12]
```diff
@@ -120,16 +120,16 @@
 
                     <td>${counter++}</td>
                     <td>
-                        <semui:listIcon type="${ie.tipp?.title?.type?.value}"/>
+                        <semui:listIcon type="${ie.tipp.title.class.name}"/>
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/surveyTitles.gsp:123]: ie.tipp.title.printTitleType()

File: app/grails-app/views/templates/tipps/_entitlementTable.gsp [hunk 13]
```diff
@@ -95,7 +54,7 @@
                                 </div>
                             </g:else>
 
-                            <semui:listIcon type="${tipp.title?.type?.value}"/>
+                            <semui:listIcon type="${ie.tipp.title.class.name}"/>
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/tipps/_entitlementTable.gsp:57]: ie.tipp.title.printTitleType()

File: app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy [hunk 14]
```diff
@@ -1218,6 +1229,63 @@ class SurveyController {
...
+            flash.error = g.message(code: "default.notAutorized.message")
+            redirect(url: request.getHeader('referer'))
+        }
+
+        result.subscriptionInstance = result.surveyConfig.subscription.getDerivedSubscriptionBySubscribers(result.participant)
+
+        result.ies = subscriptionService.getIssueEntitlementsNotFixed(result.subscriptionInstance)
+        result.iesListPriceSum = 0
+        result.ies?.each{
+            result.iesListPriceSum = result.iesListPriceSum + (it?.priceItem ? (it?.priceItem?.listPrice ? it?.priceItem?.listPrice : 0) : 0)
```
Comment (klodav) [inline app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy:1258]: Can _it_ really be _null_ ?

File: app/grails-app/views/survey/surveyParticipants.gsp (hunk 11 above)
Comment (Moetez16) [reply app/grails-app/views/survey/surveyParticipants.gsp:58]: Yes. Next Release.

File: app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy [hunk 15]
```diff
@@ -1218,6 +1229,63 @@ class SurveyController {
...
+        }
+
+
+        result.iesFix = subscriptionService.getIssueEntitlementsFixed(result.subscriptionInstance)
+        result.iesFixListPriceSum = 0
+        result.iesFix?.each{
+            result.iesFixListPriceSum = result.iesFixListPriceSum + (it?.priceItem ? (it?.priceItem?.listPrice ? it?.priceItem?.listPrice : 0) : 0)
```
Comment (klodav) [inline app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy:1265]: Can _it_ really be _null_ ?

File: app/grails-app/views/survey/surveyParticipants.gsp (hunk 11 above)
Comment (Moetez16) [reply app/grails-app/views/survey/surveyParticipants.gsp:58]: Maybe :D

File: app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy (hunk 14 above)
Comment (Moetez16) [reply app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy:1258]: Maybe :D

Comment (croqueGrec09) [reply app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy]: Yes, in the eyes of Kint who produced this legacy s**t!

File: app/grails-app/views/survey/_actions.gsp (hunk 7 above)
Comment (klodav) [reply app/grails-app/views/survey/_actions.gsp:56]: Agreed .. NEVER use ? on constants

File: app/grails-app/views/survey/surveyTitlesSubscriber.gsp [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            <div class="ui card la-time-card">
+                <div class="content">
+                    <div class="header"><g:message code="subscription.entitlement.plural"/></div>
+                </div>
+
+                <div class="content">
+                    <div class="header"><g:message code="renewEntitlementsWithSurvey.currentFixedEntitlements"/></div>
+                    <dl>
+                        <dt class="control-label">${message(code: 'myinst.selectPackages.pkg_titles')}</dt>
+                        <dd>${iesFix.size() ?: 0}</dd>
```
Comment (klodav) [inline app/grails-app/views/survey/surveyTitlesSubscriber.gsp:124]: ${iesFix.size()}

File: app/grails-app/views/survey/surveyTitlesSubscriber.gsp [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+                        <dt class="control-label">${message(code: 'tipp.listPrice')}</dt>
+                        <dd><g:formatNumber number="${iesFixListPriceSum}" type="currency"/></dd>
+                    </dl>
+                </div>
+
+                <div class="content">
+                    <div class="header"><g:message code="renewEntitlementsWithSurvey.currentEntitlements"/></div>
+                    <dl>
+                        <dt class="control-label">${message(code: 'myinst.selectPackages.pkg_titles')}</dt>
+                        <dd>${ies.size() ?: 0}</dd>
```
Comment (klodav) [inline app/grails-app/views/survey/surveyTitlesSubscriber.gsp:136]: ${ies.size()}

**Pull request merged**